// @react-three/drei/web/index.js
import 'blob-polyfill';
import '../helpers/base64.js';
import '../assets/cloud.base64.js';
import 'react';
import '@react-three/fiber';
export { AdaptiveDpr } from '../core/AdaptiveDpr.js';
export { AdaptiveEvents } from '../core/AdaptiveEvents.js';
import '@babel/runtime/helpers/esm/extends';
import '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
export { Box, Circle, Cone, Cylinder, Dodecahedron, Extrude, Icosahedron, Lathe, Octahedron, Parametric, Plane, Polyhedron, Ring, Sphere, Tetrahedron, Torus, TorusKnot, Tube } from '../core/shapes.js';
import 'react-merge-refs';
export { Billboard } from '../core/Billboard.js';
import 'three-stdlib';
export { CameraShake } from '../core/CameraShake.js';
import 'three';
export { Center } from '../core/Center.js';
export { useTexture } from '../core/useTexture.js';
export { Cloud } from '../core/Cloud.js';
export { ContactShadows } from '../core/ContactShadows.js';
export { CubeCamera } from '../core/CubeCamera.js';
export { Line } from '../core/Line.js';
export { CubicBezierLine } from '../core/CubicBezierLine.js';
export { CurveModifier } from '../core/CurveModifier.js';
export { Detailed } from '../core/Detailed.js';
export { DeviceOrientationControls } from '../core/DeviceOrientationControls.js';
export { Effects, isWebGL2Available } from '../core/Effects.js';
import 'use-asset';
import '../helpers/environment-assets.js';
export { Environment } from '../core/Environment.js';
export { FlyControls } from '../core/FlyControls.js';
export { OrthographicCamera } from '../core/OrthographicCamera.js';
export { useCamera } from '../core/useCamera.js';
export { GizmoHelper, useGizmoContext } from '../core/GizmoHelper.js';
export { GizmoViewcube } from '../core/GizmoViewcube.js';
export { GizmoViewport } from '../core/GizmoViewport.js';
export { MapControls } from '../core/MapControls.js';
import '@babel/runtime/helpers/esm/defineProperty';
export { MeshDistortMaterial } from '../core/MeshDistortMaterial.js';
export { MeshWobbleMaterial } from '../core/MeshWobbleMaterial.js';
export { OrbitControls } from '../core/OrbitControls.js';
export { PerspectiveCamera } from '../core/PerspectiveCamera.js';
export { PointerLockControls } from '../core/PointerLockControls.js';
export { PositionalAudio } from '../core/PositionalAudio.js';
export { Preload } from '../core/Preload.js';
export { QuadraticBezierLine } from '../core/QuadraticBezierLine.js';
export { MeshReflectorMaterial } from '../materials/MeshReflectorMaterial.js';
export { Reflector } from '../core/Reflector.js';
export { RoundedBox } from '../core/RoundedBox.js';
export { ScreenQuad } from '../core/ScreenQuad.js';
export { Shadow } from '../core/Shadow.js';
export { Sky, calcPosFromAngles } from '../core/Sky.js';
export { Stars } from '../core/Stars.js';
import 'three/examples/js/libs/stats.min';
import '../helpers/useEffectfulState.js';
export { Stats } from '../core/Stats.js';
import 'troika-three-text';
export { Text } from '../core/Text.js';
export { TrackballControls } from '../core/TrackballControls.js';
import 'lodash.pick';
import 'lodash.omit';
export { TransformControls } from '../core/TransformControls.js';
export { useCubeTexture } from '../core/useCubeTexture.js';
export { useFBX } from '../core/useFBX.js';
export { useGLTF } from '../core/useGLTF.js';
import 'zustand';
export { useProgress } from '../core/useProgress.js';
export { useAspect } from '../core/useAspect.js';
import 'detect-gpu';
export { useDetectGPU } from '../core/useDetectGPU.js';
export { useHelper } from '../core/useHelper.js';
export { useContextBridge } from '../core/useContextBridge.js';
export { useAnimations } from '../core/useAnimations.js';
export { useFBO } from '../core/useFBO.js';
export { useSimplification } from '../core/useSimplification.js';
export { useTessellation } from '../core/useTessellation.js';
export { useEdgeSplit } from '../core/useEdgeSplit.js';
export { shaderMaterial } from '../core/shaderMaterial.js';
export { softShadows } from '../core/softShadows.js';
import '../helpers/matcap-assets.js';
export { useMatcapTexture } from '../core/useMatcapTexture.js';
import '../helpers/normal-assets.js';
export { useNormalTexture } from '../core/useNormalTexture.js';
export { meshBounds } from '../core/meshBounds.js';
import 'react-dom';
export { Html } from './Html.js';
import '@react-spring/web';
export { Loader } from './Loader.js';
