// @react-three/drei/core/GizmoViewport.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutPropertiesLoose from '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import { createElement, useMemo, useState } from 'react';
import { CanvasTexture } from 'three';
import { useGizmoContext } from './GizmoHelper.js';

function Axis({
  color,
  rotation
}) {
  return /*#__PURE__*/createElement("group", {
    rotation: rotation
  }, /*#__PURE__*/createElement("mesh", {
    position: [0.4, 0, 0]
  }, /*#__PURE__*/createElement("boxGeometry", {
    args: [0.8, 0.05, 0.05]
  }), /*#__PURE__*/createElement("meshBasicMaterial", {
    color: color,
    toneMapped: false
  })));
}

function AxisHead(_ref) {
  let {
    arcStyle,
    label,
    labelColor
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, ["arcStyle", "label", "labelColor"]);

  const texture = useMemo(() => {
    const canvas = document.createElement('canvas');
    canvas.width = 64;
    canvas.height = 64;
    const context = canvas.getContext('2d');
    context.beginPath();
    context.arc(32, 32, 16, 0, 2 * Math.PI);
    context.closePath();
    context.fillStyle = arcStyle;
    context.fill();

    if (label) {
      context.font = '24px Arial';
      context.textAlign = 'center';
      context.fillStyle = labelColor;
      context.fillText(label, 32, 41);
    }

    return new CanvasTexture(canvas);
  }, [arcStyle, label, labelColor]);
  const [active, setActive] = useState(false);
  const scale = (label ? 1 : 0.75) * (active ? 1.2 : 1);

  const handlePointerOver = e => {
    setActive(true);
    e.stopPropagation();
  };

  const handlePointerOut = e => {
    setActive(false);
    e.stopPropagation();
  };

  return /*#__PURE__*/createElement("sprite", _extends({
    scale: [scale, scale, scale],
    onPointerOver: handlePointerOver,
    onPointerOut: handlePointerOut
  }, props), /*#__PURE__*/createElement("spriteMaterial", {
    map: texture,
    alphaTest: 0.3,
    opacity: label ? 1 : 0.75,
    toneMapped: false
  }));
}

const GizmoViewport = (_ref2) => {
  let {
    axisColors = ['#ff3653', '#8adb00', '#2c8fff'],
    labelColor = '#000'
  } = _ref2,
      props = _objectWithoutPropertiesLoose(_ref2, ["axisColors", "labelColor"]);

  const [colorX, colorY, colorZ] = axisColors;
  const {
    tweenCamera,
    raycast
  } = useGizmoContext();
  const axisHeadProps = {
    labelColor,
    onPointerDown: e => {
      tweenCamera(e.object.position);
      e.stopPropagation();
    },
    raycast
  };
  return /*#__PURE__*/createElement("group", _extends({
    scale: [40, 40, 40]
  }, props), /*#__PURE__*/createElement(Axis, {
    color: colorX,
    rotation: [0, 0, 0]
  }), /*#__PURE__*/createElement(Axis, {
    color: colorY,
    rotation: [0, 0, Math.PI / 2]
  }), /*#__PURE__*/createElement(Axis, {
    color: colorZ,
    rotation: [0, -Math.PI / 2, 0]
  }), /*#__PURE__*/createElement(AxisHead, _extends({
    arcStyle: colorX,
    position: [1, 0, 0],
    label: "X"
  }, axisHeadProps)), /*#__PURE__*/createElement(AxisHead, _extends({
    arcStyle: colorY,
    position: [0, 1, 0],
    label: "Y"
  }, axisHeadProps)), /*#__PURE__*/createElement(AxisHead, _extends({
    arcStyle: colorZ,
    position: [0, 0, 1],
    label: "Z"
  }, axisHeadProps)), /*#__PURE__*/createElement(AxisHead, _extends({
    arcStyle: colorX,
    position: [-1, 0, 0]
  }, axisHeadProps)), /*#__PURE__*/createElement(AxisHead, _extends({
    arcStyle: colorY,
    position: [0, -1, 0]
  }, axisHeadProps)), /*#__PURE__*/createElement(AxisHead, _extends({
    arcStyle: colorZ,
    position: [0, 0, -1]
  }, axisHeadProps)), /*#__PURE__*/createElement("ambientLight", {
    intensity: 0.5
  }), /*#__PURE__*/createElement("pointLight", {
    position: [10, 10, 10],
    intensity: 0.5
  }));
};

export { GizmoViewport };
