// @react-three/drei/core/Line.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutPropertiesLoose from '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import { forwardRef, useState, useMemo, useLayoutEffect, createElement } from 'react';
import { Vector2, Vector3, Color } from 'three';
import { Line2, LineMaterial, LineGeometry } from 'three-stdlib';

const Line = /*#__PURE__*/forwardRef(function Line(_ref, ref) {
  let {
    points,
    color = 'black',
    vertexColors,
    lineWidth,
    dashed
  } = _ref,
      rest = _objectWithoutPropertiesLoose(_ref, ["points", "color", "vertexColors", "lineWidth", "dashed"]);

  const [line2] = useState(() => new Line2());
  const [lineMaterial] = useState(() => new LineMaterial());
  const [resolution] = useState(() => new Vector2(512, 512));
  const lineGeom = useMemo(() => {
    const geom = new LineGeometry();
    const pValues = points.map(p => p instanceof Vector3 ? p.toArray() : p);
    geom.setPositions(pValues.flat());

    if (vertexColors) {
      const cValues = vertexColors.map(c => c instanceof Color ? c.toArray() : c);
      geom.setColors(cValues.flat());
    }

    return geom;
  }, [points, vertexColors]);
  useLayoutEffect(() => {
    line2.computeLineDistances();
  }, [points, line2]);
  useLayoutEffect(() => {
    if (dashed) {
      lineMaterial.defines.USE_DASH = '';
    } else {
      // Setting lineMaterial.defines.USE_DASH to undefined is apparently not sufficient.
      delete lineMaterial.defines.USE_DASH;
    }

    lineMaterial.needsUpdate = true;
  }, [dashed, lineMaterial]);
  return /*#__PURE__*/createElement("primitive", _extends({
    dispose: undefined,
    object: line2,
    ref: ref
  }, rest), /*#__PURE__*/createElement("primitive", {
    dispose: undefined,
    object: lineGeom,
    attach: "geometry"
  }), /*#__PURE__*/createElement("primitive", _extends({
    dispose: undefined,
    object: lineMaterial,
    attach: "material",
    color: color,
    vertexColors: Boolean(vertexColors),
    resolution: resolution,
    linewidth: lineWidth,
    dashed: dashed
  }, rest)));
});

export { Line };
